fearless/ApplicationLayer/Services/WalletBalanceFetcher.swift

```diff
@@ -0,0 +1,405 @@
+
+    private let metaAccountRepository: AnyDataProviderRepository<MetaAccountModel>
+    private let chainRepository: AnyDataProviderRepository<ChainModel>
+    private let operationQueue: OperationQueue
+    private let logger: Logger
+
+    private var completion: WalletBalanceResultCompletion?
+    private var deliverQueue: DispatchQueue?
+
+    private lazy var accountInfosAdapters: [String: AccountInfoSubscriptionAdapter] = [:]
+    private lazy var accountInfos: [ChainAssetKey: AccountInfo] = [:]
+    private lazy var chainAssets: [ChainAsset] = []
+    private lazy var metaAccounts: [MetaAccountModel] = []
+    private lazy var prices: [PriceData] = []
+
+    // MARK: - Constructor
+
+    init(
+        metaAccountRepository: AnyDataProviderRepository<MetaAccountModel>,
+        priceLocalSubscriptionFactory: PriceProviderFactoryProtocol,
+        chainRepository: AnyDataProviderRepository<ChainModel>,
+        operationQueue: OperationQueue,
+        logger: Logger
+    ) {
+        self.priceLocalSubscriptionFactory = priceLocalSubscriptionFactory
+        self.metaAccountRepository = metaAccountRepository
+        self.chainRepository = chainRepository
+        self.operationQueue = operationQueue
+        self.logger = logger
+    }
+
+    // MARK: - Public methods
+
+    func fetchBalanceFor(
+        walletId: String?,
+        deliverOn queue: DispatchQueue?,
+        completion: @escaping WalletBalanceResultCompletion
+    ) {
+        self.completion = completion
+        deliverQueue = queue
+
+        if let identifier = walletId {
+            fetchMetaAccount(by: identifier)
+        } else {
+            fetchAllMetaAccounts()
+        }
+    }
+
+    // MARK: - Private methods
+
+    private func buildBalance() {
+        guard
+            accountInfos.isNotEmpty,
+            prices.isNotEmpty
+        else {
+            return
+        }
+
+        let builderBlock = BlockOperation {
+            let walletBalances = self.walletBalanceBuilder.buildBalance(
+                accountInfos: self.accountInfos,
+                metaAccounts: self.metaAccounts,
+                chainAssets: self.chainAssets,
+                prices: self.prices
+            )
+
+            self.deliverOnCompletion(.success(walletBalances))
+        }
+
+        operationQueue.addOperation(builderBlock)
+    }
+
+    private func handle(_ wallets: [MetaAccountModel], _ chains: [ChainModel]) {
+        let chainsAssets = chains.map(\.chainAssets).reduce([], +)
+        chainAssets = chainsAssets
+        metaAccounts = wallets
+
+        subscribeToAccountInfo(for: wallets, chainsAssets)
+        subscribeToPrices(for: chains)
+    }
+
+    private func fetchMetaAccount(by identifire: String) {
+        let metaAccountOperation = metaAccountRepository.fetchOperation(
+            by: identifire,
+            options: RepositoryFetchOptions()
+        )
+        let chainsOperation = fetchChainsOperation()
```

Comment:
why do you create metaAccountOperation directly in this method, but chainsOperation from the separate method? it looks the same